File: linear_regression.cpp
#include <vector>
#include <iostream>
#include <ostream>
#include <type_traits>

namespace TypeChecks
{
    using std::begin;
    using std::end;

    template <typename T>
    concept is_iterable = requires(T &t) {
        std::begin(t) != std::end(t);
        ++std::begin(t);
        *std::begin(t);
    };

    template <typename I>
    concept is_numeric = requires(I &i) {
        { i + 0 } -> std::same_as<I>;
        { i - 0 } -> std::same_as<I>;
        { i * 1 } -> std::same_as<I>;
        { i / 1 } -> std::same_as<I>;
        { ++i   } -> std::same_as<I &>;
        { i++   } -> std::same_as<I>; 
    };
}

namespace ML
{
    typedef std::vector<float> floatVector;

    enum Ops {
        ADD,
        SUB,
        MULTIPLY,
        DIVIDE
    };

    // Y = M * X + C
    class LinearRegression {

    public:

        floatVector inputs;
        floatVector outputs;
        float slope;
        float intercept;
        float learningRate;

        LinearRegression(
            floatVector i,
            floatVector o,
            float m = 1,
            float c = 1,
            float lr = 0.001
        ) : inputs(i), outputs(o), slope(m), intercept(c), learningRate(lr) {}


        float operate(float v1, float v2, Ops op){
            switch (op) 
            {
                case Ops::ADD:
                    return v1 + v2;
                case Ops::SUB:
                    return v1 - v2;
                case Ops::MULTIPLY:
                    return v1 * v2;
                case Ops::DIVIDE:
                    if (v2 == 0.0f) throw "Division by zero !";
                    return v1 / v2;
                default:
                    throw "Unkonwn operation passed !";
            }
        }

        template <typename T> // val can be a number or a vector
        floatVector operateOnVectorValuesWith(const T &val, const floatVector &mainVector, Ops operation) 
        {
            floatVector res;
            // Scaler multiplication
            if constexpr (std::is_arithmetic<T>::value) {    
                for (auto ele : mainVector) {
                    result.push_back(operate(ele, val, operation));
                }
            }
            // Vector multiplication
            else 
            { 
                static_assert(
                    std::is_same<T, std::vector<float>>::value, 
                    "val must be either a number or a vector of floats"
                );

                if (val.size() == mainVector.size()) {
                    for (size_t i = 0; i < mainVector.size(); ++i) {
                        result.push_back(operate(mainVector[i], val[i], operation)); //mainVector[i] * val[i]
                    }
                } 
                else {
                    throw std::invalid_argument("Vector lengths don't match!");
                }
            }

            return res;
        }

        float mean(floatVector vec){
            float sum = 0.0;
            for(auto v: vec){
                sum += v;
            }
            return sum / vec.size();
        }

        floatVector predict(float sl=0.0, float in=0.0){
            auto slope = (sl == 0.0) ? this->slope : sl;
            auto intercept = (in == 0.0) ? this->intercept : in;
            auto temp = operateOnVectorValuesWith(intercept, inputs, Ops::MULTIPLY);
            return operateOnVectorValuesWith(slope, temp, Ops::ADD);
        }

        float getMSELoss(floatVector predictions){
            auto temp = operateOnVectorValuesWith(outputs, predictions, Ops::SUB);
            auto tempSq = operateOnVectorValuesWith(temp, temp, Ops::MULTIPLY);
            return mean(tempSq);
        }

        void backwardPropagationBGD(floatVector predictions){
            auto diff = operateOnVectorValuesWith(predictions, outputs, Ops::SUB);
            auto temp = operateOnVectorValuesWith(inputs, diff, Ops::MULTIPLY);
            float gradientSlope = 2 * mean(temp);
            float gradientIntercept = 2 * mean(diff);
            slope -= learningRate * gradientSlope;
            intercept -= learningRate * gradientIntercept;
        }

        void train(int epochs){
            for(int i = 0; i < epochs; i++){
                auto predictions = predict();
                backwardPropagationBGD(predictions);
                std::cout << "EPOCH: " << i << ", MSE Loss: " << getMSELoss(predictions) << std::endl;
            }
        }
        
    };
};

File: linear_regression.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.axes as ax
from matplotlib.animation import FuncAnimation

class LinearRegression:

    def __init__(self, dependentVariables, outputs, slope=1, intercept=1, learningRate=0.0001) -> None:
        self.slope = slope
        self.intercept = intercept
        self.dependentVars = dependentVariables
        self.outputs = outputs
        self.lr = learningRate
        self.samples = len(outputs)

        print(slope, intercept, self.lr, self.samples)

    def forwardPropagation(self):
        return self.slope + np.multiply(self.intercept, self.dependentVars)

    # MEAN SQUARED ERROR
    def getMSELoss(self, predictions): 
        return np.mean((self.outputs - predictions) ** 2)
    
    # MEAN ABSOLUTE ERROR 
    def getMAELoss(self, predictions):
        return np.mean(np.abs(self.outputs - predictions))
    
    # ROOT MEAN SQUARED ERROR
    def getRMSELoss(self, predictions):
        return np.sqrt(self.getMSELoss(predictions))
    
    # BATCH GRADIENT DECENT
    def backwardPropagationBGD(self, predictions):
        diff = predictions - self.outputs
        # Actual formula = 2 / n * summation of ( diff * inputs )
        gradient_slope = 2 * (np.mean(np.multiply(self.dependentVars, diff)))
        # Actual formula = 2 / n * summation of (ypred - yreal) i.e diff
        gradient_intercept = 2 * (np.mean(predictions - self.outputs))

        self.slope -= self.lr * gradient_slope
        self.intercept -= self.lr * gradient_intercept

    def train(self, epochs=10):
        for epoch in range(epochs):
            predictions = self.forwardPropagation()
            self.backwardPropagationBGD(predictions)
            
            # Calculate and print the loss at each epoch
            mse = self.getMSELoss(predictions)
            mae = self.getMAELoss(predictions)
            rmse = self.getRMSELoss(predictions)

            if mse > 1e10:  # You can adjust this threshold
                print("Loss is too high. Stopping training.")
                break
            
        print(f"\n Epochs: {epoch+1}/{epochs} | MSE={mse} | MAE={mae} | RMSE={rmse} \n")

    def predict(self, new_data):
        return self.slope + np.multiply(self.intercept, new_data)

    def plot(self, newData, newDataOutput):
        plt.figure(figsize=(10, 6))
        plt.scatter(newData, newDataOutput, color='blue', label='Actual Data')

        x_values = np.linspace(min(self.dependentVars), max(self.dependentVars), 100)
        y_values = self.slope + self.intercept * x_values
        plt.plot(x_values, y_values, color='red', label='Regression Line')

        predicted_outputs = self.predict(newData)
        plt.scatter(newData, predicted_outputs, color='green', label='Predicted Data', marker='x')
        
        plt.xlabel('Dependent Variables')
        plt.ylabel('Outputs')
        plt.title('Linear Regression Fit with Predictions')
        plt.legend()
        plt.show()


url = 'https://media.geeksforgeeks.org/wp-content/uploads/20240320114716/data_for_lr.csv'
data = pd.read_csv(url)

# Drop the missing values
data = data.dropna()

# print(data)

# training dataset and labels
train_input = np.array(data.x[0:5]).reshape(5, 1)
train_output = np.array(data.y[0:5]).reshape(5, 1)
test_input = np.array(data.x[199:204]).reshape(5, 1)
test_output = np.array(data.y[199:204]).reshape(5, 1)

lr = LinearRegression(train_input, train_output)
lr.train(500)
res = lr.predict(test_input)
lr.plot(test_input, test_output)

for i in range(len(res)):
    print("X = ", test_input[i], " | Predicted = ", res[i], " | Real =", test_output[i]);